### PR TITLE
[SPE-823] Add logic to pull cert on startup or kick off acme order if doesn't exist

### PR DESCRIPTION
**Pull request by donaltuohy**
# Why
Need to share cert across cages so on startup we should either:
* pull cert from S3
* If it doesn't exist, order it and write to S3.

# How
Based off of #75 so that should be merged first.

The flow is:
* Read the cert from S3
    * If it does exist, decrypt and use it.
    * If it doesn't exist, check for a lock:
       * If the lock does exist:
          * Check if lock has expired:
             * It hasn't expired, wait 3 seconds and then run this loop again to check
               * It has expired:
                  * Write lock and check persisted
                  * Kick off order for zerossl and encrypt the cert with an e3 call.
                  * Upload encrypted certificate to S3.

There's a loop that will run in the data plane on startup that follows this flow:

ACME ORDERING

### CODE REVIEW

File: control-plane/src/config_server.rs [hunk 1]
```diff
@@ -418,13 +426,17 @@ async fn get_acme_jwk(acme_account_details: AcmeAccountDetails) -> ServerResult<
         .map_err(ServerError::HyperHttp)
 }
 
-fn validate_order_identifiers(
+fn valid_order_identifiers(
     payload: NewOrderPayload,
     cage_context: configuration::CageContext,
 ) -> bool {
+    let underscored_app_uuid = cage_context.app_uuid.replace('-', "_");
```
Comment (lfarrel6) [inline control-plane/src/config_server.rs]: can use the context function here:
```suggestion
    let underscored_app_uuid = cage_context.underscored_app_uuid();
```

File: control-plane/src/configuration.rs [hunk 2]
```diff
@@ -77,8 +77,12 @@ impl CageContext {
         }
     }
 
+    pub fn underscored_app_uuid(&self) -> String {
+        self.app_uuid.replace('-', "_")
+    }
+
     pub fn get_namespace_string(self) -> String {
-        format!("{}/{}/{}", self.team_uuid, self.app_uuid, self.cage_uuid)
+        format!("{}/{}", self.underscored_app_uuid(), self.cage_name)
```
Comment (lfarrel6) [inline control-plane/src/configuration.rs]: this shouldn't be consuming self

File: control-plane/src/main.rs [hunk 3]
```diff
@@ -38,12 +38,14 @@ async fn main() -> Result<()> {
         configuration::get_cert_provisoner_host(),
         3000,
         shared::ENCLAVE_CERT_PORT,
+        true,
```
Comment (lfarrel6) [inline control-plane/src/main.rs:41]: this feels a bit too easy to change accidentally/without thinking. Wdyt about having two constructors? `with_internal_dns` and `with_external_dns`?

The ideal might be to have types for the two DNS services, and have constructors bound to internal/external.

Comment (lfarrel6) [reply control-plane/src/main.rs:41]: Not sure creating the DNS service in the constructor makes great sense when there's two options.

File: data-plane/src/acme/cert.rs [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+                        None => {
+                            println!("[ACME] Lock not found, creating lock, creating Certificate, encrypting Certificate, persisting Certificate, deleting lock");
+                            //Lock doesn't exist - create lock - create Certificate - encrypt Certificate - persist Certificate - delete lock
+                            let unencrypted_certificate_maybe = self
+                                .create_certificate_and_persist_with_lock(key.clone())
+                                .await?;
+                            persisted_certificate = unencrypted_certificate_maybe;
+                        }
+                    }
+                }
```
Comment (lfarrel6) [inline data-plane/src/acme/cert.rs:166]: Wdyt about splitting out the nesting here? Either into earlier returns or isolated functions? The first arm of this match (cert found in storage) doesn't need to be a match arm anyway.

Comment (lfarrel6) [reply data-plane/src/acme/cert.rs:166]: Also, any intermittent S3 error here seems to be a short-circuit/bubble up. Is that intentional?

**Review by lfarrel6 — CHANGES_REQUESTED** [review]
Couple Qs/suggestions

**Review by hanneary — APPROVED** [review]

File: data-plane/src/acme/cert.rs (hunk 4 above)
Comment (donaltuohy) [reply data-plane/src/acme/cert.rs]: Yeah good shout, it was a bit of a beast. I've broken it down so the possible paths are clearer. I also added retries to the s3 calls with backoff and jitter.

**Review by lfarrel6 — APPROVED** [review]

**Pull request merged**